components/raftstore/src/store/peer.rs

```diff
@@ -4513,33 +4570,133 @@ where
+            matches!(peer_disk_usage, DiskUsage::Normal))
         {
             info!(
                 "reject transferring leader";
                 "region_id" => self.region_id,
                 "peer_id" => self.peer.get_id(),
-                "from" => from,
+                "from" => msg.get_from(),
                 "pending_snapshot" => pending_snapshot,
                 "disk_usage" => ?ctx.self_disk_usage,
             );
-            return;
+            return true;
+        }
+        false
+    }
+
+    /// Before ack the transfer leader message sent by the leader.
+    /// For easy understanding, this stage is called `PreBecomeLeaderState`.
+    /// Currently, it only warms up the entry cache in this stage.
+    ///
+    /// This return a tuple: (should_ack, is_first_time).
+    /// * `should_ack` means that the the TransferLeaderMsg should be acked
+    ///   immediately. When cache is warmed up or the warmup operation is
+    ///   timeout, it is true.
+    /// * `is_first_time` implies whether it isthe first time entering this
+    ///   state.
+    pub fn pre_ack_transfer_leader_msg<T>(
+        &mut self,
+        ctx: &mut PollContext<EK, ER, T>,
+        msg: &eraftpb::Message,
+    ) -> (bool, bool) {
+        // The start index of warmup range is leader's entry_cache_start_index,
+        // and it is equal to the lowest matched index in general.
+        let warmup_range_start = msg.get_index();
+        let last_index = self.get_store().last_index();
+        let mut should_ack_now = false;
+        let mut low = warmup_range_start;
+
+        // Need not to warm up when the index is 0.
+        // There are two cases where index can be 0:
+        // 1. During rolling upgrade, old instances may not support warmup.
+        // 2. The leader's entry cache is empty.
+        if warmup_range_start == 0 {
+            WARM_UP_ENTRY_CACHE_COUNTER.no_need.inc();
+            should_ack_now = true;
+        } else if warmup_range_start > last_index {
+            // warmup range start should not be large than the last index.
+            // Since the index is leader's entry_cache_first_index, so this is
+            // unlikely to happen.
+            WARM_UP_ENTRY_CACHE_COUNTER.invalid_index.inc();
+            warn!(
+                "the start index of warmup range is too large";
+                "region_id" => self.region_id,
+                "peer_id" => self.peer.get_id(),
+                "warmup_range_start" => warmup_range_start,
+                "last_index" => last_index,
+            );
+            should_ack_now = true;
+        } else {
+            if warmup_range_start < self.last_compacted_idx {
+                low = self.last_compacted_idx
+            };
+            // Check if the entry cache is already warmed up.
+            if let Some(first_index) = self.get_store().entry_cache_first_index() {
+                if low >= first_index {
+                    WARM_UP_ENTRY_CACHE_COUNTER.already.inc();
```

Comment:
I think the metrics is too verbose..what's the purpose of "already" counter?

Reply:
fixed